src/Catalyst.Core.Modules.Web3/Controllers/Handlers/EthGetBlockByNumberHandler.cs

```diff
@@ -22,42 +22,91 @@
 #endregion
 
 using System;
+using System.Buffers.Binary;
+using Catalyst.Abstractions.Consensus.Deltas;
+using Catalyst.Abstractions.Kvm;
 using Catalyst.Abstractions.Kvm.Models;
 using Catalyst.Abstractions.Ledger;
+using Google.Protobuf;
+using LibP2P;
 using Nethermind.Core;
 using Nethermind.Core.Crypto;
 using Nethermind.Dirichlet.Numerics;
+using Address = Nethermind.Core.Address;
 
 namespace Catalyst.Core.Modules.Web3.Controllers.Handlers
 {
     [EthWeb3RequestHandler("eth", "getBlockByNumber")]
     public class EthGetBlockByNumberHandler : EthWeb3RequestHandler<BlockParameter, BlockForRpc>
     {
-        protected override BlockForRpc Handle(BlockParameter address, IWeb3EthApi api)
+        protected override BlockForRpc Handle(BlockParameter block, IWeb3EthApi api)
         {
-            // this needs some kind of delta repository that knows current delta and is also able to retrieve past deltas
-            throw new NotImplementedException();
+            Cid deltaHash;
+            long blockNumber;
+
+            IDeltaCache deltaCache = api.DeltaCache;
+            IDeltaResolver deltaResolver = api.DeltaResolver;
+
+            switch (block.Type)
+            {
+                case BlockParameterType.Earliest:
+                    deltaHash = deltaCache.GenesisHash;
+                    blockNumber = 0;
+                    break;
+                case BlockParameterType.Latest:
+                    deltaHash = deltaResolver.LatestDelta;
+                    blockNumber = deltaResolver.LatestDeltaNumber;
+                    break;
+                case BlockParameterType.Pending:
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Either merge this case with the identical one on line 54 or change one of the implementations.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4872747)